NUDelta/ce-platform
# Added a new Half-Half Photo experience

### Summary of Changes
This PR introduces several core features that supported me in building the Half Half Photo Experiences

_Changes for Half Half Photo OCE_
- New [Cordova Plugin that supports a camera preview](https://github.com/cordova-plugin-camera-preview/cordova-plugin-camera-preview) that can be interacted with HTML
- halfhalfParticipate Template [api_custom.js], with lots of additional helpers, events, and functions that support taking photos with the new Cordova plugin. The main way it does this is by processing base64 images, and storing the contents in an `<img src="base64..."`. The uploading of image data uses the same logic as was there before.
- halfhalfResults Template [api_custom_results.js], with its associated helpers/functions for 
- New subscriptions of data for `api.custom` [router.js], since this is the first time a participate template needs data that was submitted previously. In Half Half Travel, the 1st half is displayed while the person is submitting the second half of the photo.  We might want to think about template organization, given that api_custom and api_custom_results may take on similar forms.
- New subscriptions for users data for `api.customresults` [router.js], since we need it to display user information in the results template.
- New OCES created for halfhalf written in `testingconstants.js`.
- Template helper to display the full image (since most of our templates were showing a reduced size square "thumb", which was not desirable for halfhalf photos)
- Template styles (S/CSS) needed to make the background transparent, so that the CameraPreview could be shown behind the HTML.

_Fixes to Cerebro Server_:
- Progressor [progressor.js] now imports methods like `notify` and `adminUpdates` [progressorHelpers.js] which were being called in this file, but were failing before here. Now, people are notified after an experience is finished (`notify`) and people's participation now is recorded in the appropriate places (`adminUpdates`) 
- Updated some tests that had outdated detectors. TODO: these end-to-end tests are not being used effectively right now, or really at all in a systematic way to guard against bugs in new PRs.
- Remove Images in the DB too when doing a `freshDatabase` Meteor methods call

_Additional support for General CE Development_
- Discovered how to deploy enterprise apps more easily [ipaHelper.sh]
- Discovered how to do iOS development running a local server and documented the setup [README.md]
- better Meteor method `createOCE` [createNewExperiences.js] which will allow us to create new experiences that are defined in CONSTANTS.EXPERIENCES.  It is a more general version of what was there before 
- In order to test GPS data with a local server, background Geolocation will POST the GPS data to a non secure (http vs https) route [location_tracking.js].  This setting still allows GPS data to be posted to the heroku server too.

### Tests

#### Testing with Local Deploy
Half Half Night or Half Half Day are good experiences to setup so Half Half OCE always comes up, and we can test the interaction flow.  Additionally, you can use the detectors included in Half Half Asynch / Synch (grocery, commuting) and use the simulatelocation.py script to move users to appropriate locations

Testing halfhalf Participate and Results Templates

0. Go to testingconstants.js and only include the experiences you want loaded in. at the bottom of the file were `CONSTANTS` is defined, you can do something like this ```'EXPERIENCES': (({ halfhalfAsynch }) => ({ halfhalfAsynch}))(EXPERIENCES),```
1. Start local development server
2. _If testing with a web app_: Use the simulatelocation.py command to move the test users locations.
2. _If testing with the mobile app_: Login as any of the test users. Then quit the app (to register the login with background location tracking). Now open app and login again.  Because location tracking is now enabled, you should have received an experience. 
3. A half half experience should be triggered, and push notifications will be sent to some userId
4. _Only if testing with a web app_: Start up `meteor mongo` and find all test users and their uids `db.users.find().pretty()`
5. _Only if testing with a web app_: Ctrl+F in terminal to find the UID; now that you have the name, login as that person on ce-platform app (Web or Mobile)
6. Go to Home or Refresh, the Experience should be there. Participate!
7. Test what you need in the Participate UI!  Submit 1st photo, either by using camera, or using a dummy photo by clicking the glyphicon-paper clip button. (__note__: submitting photos from mobile app to local server takes hella long, i.e. its never worked for me ever.  If I'm testing locally, I submitted dummy photos from the web browser)
8. Should be directed to results template
9. Start a new user session, either by logging out on mobile or by opening up a new tab of web browser and opening the app.
10. Repeat step 2 to update the locations again.
11. A new user should have been sent push notifications.  Repeat 3 - 5 to get that new user to log in as them.
12. Repeat 6 - 8 for 2nd photo submission.
13. See if the halfhalf combined photo shows up for both in the results template.

Testing Other Experiences - Testing other experiences that use the photo upload camera module is worthwhile because the new half half Camera is using similar classes like `fileinput` and `glyphicon`
